scrapy_autoextract/providers.py

```diff
@@ -45,17 +46,56 @@ def __init__(
         self.stats = stats
         self.settings = settings
 
+    def __before__(self):
+        providers = self.request.meta.get("_autoextract_providers", [])
+        self.request.meta["_autoextract_providers"] = providers + [self]
+
     async def __call__(self):
         """Make an AutoExtract request and build a Page Input of provided class
         based on API response data.
         """
-        page_type = self.get_page_type()
-        self.stats.inc_value(f"autoextract/{page_type}/total")
+        data = await self._fetch_data()
+        if "error" in data:
+            self.stats.inc_value("autoextract/error/query")
+            raise QueryError(data["query"], data["error"])
+
+        return self.provided_class(data=data)
+
+    @classmethod
+    def register(cls):
```

Comment:
What's the advantage of this behavior over `@provides`?

Reply:
I think that's not having to duplicate the self.provided_class name.